binzx/otomi

```diff
@@ -0,0 +1,210 @@
+OTOMI_CALLER_COMMAND=${BASH_SOURCE[0]##*/}
+OTOMI_IN_DOCKER=true
+EOF
+
+helm_config="$HOME/.config/helm"
+uname -a | grep -i darwin >/dev/null && helm_config="$HOME/Library/Preferences/helm"
+
+stack_dir='/home/app/stack'
+executable="node --experimental-specifier-resolution=node ${stack_dir}/dist/otomi.js --"
+
+function mount_with_named_volume() {
+  source=$1
+  dest=$2
+
+  if [[ $(uname -s) == Darwin*  ]]; then
+    echo "$source:$dest"
+    return 0
+  fi
+
+  [ -z "$NAMED_VOLUME" ] && export NAMED_VOLUME='otomi-volume'
+  helper_container_name="otomi-volume-helper"
+
+  [ -z "$source" ] && echo "Error in function 'mount_with_named_volume': source path is empty" 1>&2
+
+  { # Create volume and copy over contents
+    docker container inspect $helper_container_name && docker rm $helper_container_name
+    docker container create --name $helper_container_name -v "$NAMED_VOLUME:$dest" hello-world &&
+      cd "$source" && docker cp . "$helper_container_name:$dest" &&
+      docker rm $helper_container_name
+  } >/dev/null 2>&1
+
+  echo "$NAMED_VOLUME:$dest"
+}
+
+stack_volume=""
+tmp_volume=""
+if [[ $base_dir == *"otomi-core"* ]] || [[ $(pwd) == *"otomi-core"* ]]; then
+  if [[ $base_dir == *"otomi-core"* ]]; then
+    stack_dir=$(cd "$base_dir/.." && pwd)
+  elif [[ $(pwd) == *"otomi-core"* ]]; then
+    stack_dir=$(pwd | awk -F'otomi-core' '{print $1"otomi-core"}')
+  fi
+
+  stack_volume="-v $(mount_with_named_volume "$stack_dir" "$stack_dir")"
+  tmp_volume="-v $(mount_with_named_volume /tmp /tmp)"
+
+  if [ -f "${stack_dir}/dist/otomi.js" ]; then # Use development distribution if exists
+    executable="node --experimental-specifier-resolution=node ${stack_dir}/dist/otomi.js --"
+  fi
+  echo "OTOMI_DEV=true" >>$tmp_env
+fi
+
+cmd="${executable} $*"
+if [ "$1" = "bash" ] && [ "$#" = "1" ]; then # If command is "otomi bash"
+  cmd="bash"
+fi
+
+pull_output=$(docker pull $otomi_tools_image 2>&1 >/dev/null)
+status=$?
+if [ "$status" -ne 0 ]; then
+  echo "$pull_output"
+  echo "Could not find an image for the supplied OTOMI_TAG: '${OTOMI_TAG}'"
+  exit $status
+fi
+
+docker run --network host --rm -it \
+  $stack_volume \
+  $tmp_volume \
+  -v $(mount_with_named_volume "${KUBECONFIG:-$HOME/.kube}" /home/app/.kube) \
+  -v $(mount_with_named_volume "$HOME/.ssh" /home/app/.ssh) \
+  -v $(mount_with_named_volume "$helm_config" /home/app/.config/helm) \
+  -v $(mount_with_named_volume "$HOME/.config/gcloud" /home/app/.config/gcloud) \
+  -v $(mount_with_named_volume "$HOME/.aws" /home/app/.aws) \
+  -v $(mount_with_named_volume "$HOME/.azure" /home/app/.azure) \
+  -v $(mount_with_named_volume "$ENV_DIR" "$stack_dir/env") \
+  -v $(mount_with_named_volume "$ENV_DIR" "$ENV_DIR") \
+  --env-file "$tmp_env" \
+  -w "$stack_dir" \
+  "$otomi_tools_image" \
+  bash -c "$cmd"
+
+status=$?
+rm -f "$tmp_env"
+
+# Named volume stuff
+docker volume rm $NAMED_VOLUME >/dev/null 2>&1
+docker rmi hello-world >/dev/null 2>&1
+
+exit $status
```

Comment:
```suggestion
npm run compile && docker run --network host --rm -it --user root:root \
  $stack_volume \
  -v ${KUBECONFIG:-$HOME/.kube/config}:/home/app/.kube/config \
  -v $HOME/.ssh:/home/app/.ssh \
  -v $helm_config:/home/app/.config/helm \
  -v $HOME/.config/gcloud:/home/app/.config/gcloud \
  -v $HOME/.aws:/home/app/.aws \
  -v $HOME/.azure:/home/app/.azure \
  -v ${GOOGLE_APPLICATION_CREDENTIALS}:${GOOGLE_APPLICATION_CREDENTIALS} \
  -v $ENV_DIR:$stack_dir/env \
  -v $ENV_DIR:$ENV_DIR \
  --env-file $tmp_env \
  -w $stack_dir \
  $otomi_tools_image \
  bash -c "$cmd"
status=$?
rm -f "$tmp_env"
exit $status
```

It was confusing that I needed to compile the code first/that the code isn't automatically compiled...

Reply:
Not doing this, as this is only a problem when running in core. Otherwise it uses the compiled version supplied by the container.

Reply:
I wonder why you would hesitate to add it, as it improves my development workflow?

Reply:
You can open a new terminal window and run: `npm run compile:watch` it will run in the background and automagically compile any TS changes.
By adding it in here, you force compilation on every execution, which is undesirable when used by a customer/user